hbase-server/src/main/java/org/apache/hadoop/hbase/master/http/MasterHbckServlet.java

```diff
@@ -0,0 +1,144 @@
+  public static final String ORPHAN_REGIONS_ON_RS = "orphan_regions_on_rs";
+  public static final String ORPHAN_REGIONS_ON_FS = "orphan_regions_on_fs";
+  public static final String HOLES = "holes";
+  public static final String OVERLAPS = "overlaps";
+  public static final String UNKNOWN_SERVERS = "unknown_servers";
+  public static final String EMPTY_REGIONINFO = "empty_regioninfo";
+
+  private static final long serialVersionUID = 1L;
+  private static final Logger LOG = LoggerFactory.getLogger(MasterHbckServlet.class);
+  private static final Gson GSON = GsonFactory.buildGson();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+    throws ServletException, IOException {
+    final HMaster master = (HMaster) getServletContext().getAttribute(HMaster.MASTER);
+    if (!master.isInitialized()) {
+      LOG.warn("Master is not initialized yet");
+      sendError(response, HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+        "master is not initialized yet");
+      return;
+    }
+    final HbckChore hbckChore = master.getHbckChore();
+    if (hbckChore == null || hbckChore.isDisabled()) {
+      LOG.warn("Hbck chore is disabled");
+      sendError(response, HttpServletResponse.SC_SERVICE_UNAVAILABLE, "Hbck chore is disabled");
+      return;
+    }
+    if (!Boolean.parseBoolean(request.getParameter("cache"))) {
+      try {
+        master.getMasterRpcServices().runHbckChore(null, null);
+      } catch (org.apache.hbase.thirdparty.com.google.protobuf.ServiceException se) {
+        LOG.warn("Failed generating a new hbck chore report; using cache", se);
+      }
+      try {
+        master.getMasterRpcServices().runCatalogScan(null, null);
+      } catch (org.apache.hbase.thirdparty.com.google.protobuf.ServiceException se) {
+        LOG.warn("Failed generating a new catalogjanitor report; using cache", se);
+      }
+    }
+    Map<String, Object> result = new HashMap<>();
+    result.put(START_TIMESTAMP, hbckChore.getCheckingStartTimestamp());
+    result.put(END_TIMESTAMP, hbckChore.getCheckingEndTimestamp());
+    final Map<String, Pair<ServerName, List<ServerName>>> inconsistentRegions =
+      hbckChore.getInconsistentRegions();
+    if (inconsistentRegions != null && !inconsistentRegions.isEmpty()) {
+      result.put(INCONSISTENT_REGIONS, inconsistentRegions);
+    }
+    final Map<String, ServerName> orphanRegionsOnRS = hbckChore.getOrphanRegionsOnRS();
+    if (orphanRegionsOnRS != null && !orphanRegionsOnRS.isEmpty()) {
+      result.put(ORPHAN_REGIONS_ON_RS, orphanRegionsOnRS);
+    }
+    final Map<String, Path> orphanRegionsOnFS = hbckChore.getOrphanRegionsOnFS();
+    if (orphanRegionsOnFS != null && !orphanRegionsOnFS.isEmpty()) {
+      result.put(ORPHAN_REGIONS_ON_FS, orphanRegionsOnFS);
+    }
+    final CatalogJanitor janitor = master.getCatalogJanitor();
+    if (janitor != null) {
+      final Report report = janitor.getLastReport();
```

Comment:
CatalogueJanitor exposes the report concept. Brilliant!

Reply:
Yes, and unfortunately the result is a mix of CatalogJanitor's nicely structured Report and HbckChore's other return types. 

Unifying these things is out of scope, though. Or, we can make it a precondition, but I won't pursue it at this time.

Reply:
Down review I accepted your proposal to implement this with Jersey, so a clean unification of these two data sources can be done when working up the model (POJO) of the report encoding to return from the request handler.